engine/src/main/java/org/hibernate/validator/internal/metadata/core/MetaConstraint.java

```diff
@@ -236,22 +235,53 @@ public boolean isSuccess() {
 		}
 	}
 
-	static final class ValueExtractionPathNode {
+	static final class TypeParameterAndExtractor {
 
 		private final TypeVariable<?> typeParameter;
 		private final ValueExtractorDescriptor valueExtractorDescriptor;
 
-		private ValueExtractionPathNode(TypeVariable<?> typeParameter, ValueExtractorDescriptor valueExtractorDescriptor) {
+		private TypeParameterAndExtractor(TypeVariable<?> typeParameter, ValueExtractorDescriptor valueExtractorDescriptor) {
 			this.typeParameter = typeParameter;
 			this.valueExtractorDescriptor = valueExtractorDescriptor;
 		}
 
-		static ValueExtractionPathNode of(ValueExtractorDescriptor valueExtractorDescriptor) {
-			return new ValueExtractionPathNode( null, valueExtractorDescriptor );
+		static TypeParameterAndExtractor of(ValueExtractorDescriptor valueExtractorDescriptor) {
+			return new TypeParameterAndExtractor( null, valueExtractorDescriptor );
+		}
+
+		static TypeParameterAndExtractor of(TypeVariable<?> typeParameter, ValueExtractorDescriptor valueExtractorDescriptor) {
+			return new TypeParameterAndExtractor( typeParameter, valueExtractorDescriptor );
+		}
+
+		@Override
+		public String toString() {
+			return "TypeParameterAndExtractor [typeParameter=" + typeParameter + ", valueExtractorDescriptor=" + valueExtractorDescriptor + "]";
+		}
+	}
+
+	private static final class ValueExtractionPathNode {
```

Comment:
As discussed on HipChat, it would be nice to have a specific implementation for the common case where we have only one node.

Other than that, it looks good to me.